Add external images processing functionality

gazal2708 opened the pull request:
- Fixes #936
- Add processExternalImages step to extract width/height from URL params
- Set hardcoded width=750 replacing existing height and width params in image url for external images
- Add width/height attributes to processed images
- Add comprehensive test coverage for external images processing
- Include test fixtures for external images markdown and expected HTML


This change mirror query-string (width, height) dimensions onto <img> and also persist other modifiers (such as quality, dpr, preset, smartcrop and other modifiers for image transformations) when authors provide them

Please ensure your pull request adheres to the following guidelines:
- [ ] make sure to link the related issues in this description
- [ ] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues


Thanks for contributing!

gazal2708 committed:
Add external images processing functionality

- Add processExternalImages step to extract width/height from URL params
- Set hardcoded width=750 and quality=65 for external images
- Add width/height attributes to processed images
- Add comprehensive test coverage for external images processing
- Include test fixtures for external images markdown and expected HTML

gazal2708 committed:
Update external images test to use Adobe delivery URLs

- Replace generic external URLs with realistic Adobe delivery service URLs
- Use format: https://delivery-p12345-e67890.adobeaemcloud.com/adobe/assets/urn:aaid:aem:...
- Remove unnecessary 'other=param' from test URLs
- Align test format with existing codebase patterns

gazal2708 committed:
Integrate external image tests into existing images test

- Add external image test cases to images.md and images.html
- Include scenarios: with width/height, with only width, without width/height
- Use Adobe delivery URLs for realistic testing
- Remove separate external-images test files
- Update rendering.test.js to use images test instead of external-images
- All tests passing with 95.23% coverage for process-external-images.js

gazal2708 committed:
Add constants for external image processing values

- Add EXTERNAL_IMAGE_WIDTH and EXTERNAL_IMAGE_QUALITY constants
- Replace hardcoded '750' and '65' values with constants
- Follow same pattern as BREAK_POINTS in create-pictures.js
- Improve code maintainability and consistency

gazal2708 committed:
Refactor constants to object pattern like BREAK_POINTS

rofe commented on src/steps/process-external-images.js:
```diff
@@ -0,0 +1,89 @@
+    return false;
+  }
+
+  const { src } = node.properties;
+  return !src.startsWith('./media_');
```
@gazal2708 yes, that may be the case today, but it just seems too broad to assume _any_ external image is from assets. Maybe there will be other/3rd party integrations in the future. It seems prudent to double check for this specific use case.

gazal2708 replied:
I have added condition in code to ensure that image src should not contain "media" urls and also should start with configured externalImageUrlPrefixes. Additionally have removed the hardcoded setting of quality param in url. Now that we will handle quality param as part of AS side code. 👍 cc: @rofe @anuraggupta228

gazal2708 committed:
Refactor external image processing and add edge case tests

- Remove hardcoded quality, preserve existing quality in URLs
- Always process external images, add attributes only if width/height present
- Add configurable externalImageUrlPrefixes support
- Add edge case tests for height-only and different domain images

gazal2708 committed:
removing comment

gazal2708 commented on src/steps/process-external-images.js:
```diff
@@ -0,0 +1,89 @@
+    const height = searchParams.get('height');
+
+    if (!width && !height) {
+      return null;
```
pushed changes regarding same

gazal2708 commented on src/steps/process-external-images.js:
```diff
@@ -0,0 +1,89 @@
+  return !src.startsWith('./media_');
+}
+
+function processExternalImage(src, alt = '', title = undefined) {
```
One clarification needed - are we referring to formal instructions doc link to be added here?

anuraggupta228 replied:
A plain English describing the function purpose.

gazal2708 committed:
removing comment

gazal2708 committed:
refactoring code for explicitly handling title

gazal2708 commented on src/steps/process-external-images.js:
```diff
@@ -0,0 +1,92 @@
+      }
+    }
+
+    if (title && title !== alt) {
```
We should not consider title in img as it adds tooltip information and for data-title, it should only be propagated with new information only when title is different from alt text. Same design being followed in existing create-pictures.js for media urls.

tripodsan commented:
@gazal2708  please sign the Adobe CLA

tripodsan commented on src/steps/process-external-images.js:
```diff
@@ -0,0 +1,93 @@
+  const { content } = state;
+  const { hast } = content;
+
+  visitParents(hast, (node) => isExternalImage(node, state), (img, parents) => {
```
if you use `visit` instead of `visitParents`, it's more efficient to replace nodes.

gazal2708 replied:
Thats a good suggestion, I have incorporated this, thankyou!

tripodsan commented on src/steps/process-external-images.js:
```diff
@@ -0,0 +1,93 @@
+    searchParams.delete('height');
+
+    searchParams.set('width', EXTERNAL_IMAGE_CONFIG.width);
+    const newSrc = `${url.origin}${url.pathname}${url.search}${url.hash}`;
```
why no `url.href` ?

gazal2708 replied:
refactored the code for incorporating this.

gazal2708 committed:
Refactor process-external-images to use visit instead of visitParents

- Switch from visitParents to visit for better performance
- Use Object.assign for direct node property modification
- Maintain same functionality with cleaner, more efficient code
- All tests passing with 97.77% coverage

gazal2708 commented on src/steps/process-external-images.js:
```diff
@@ -0,0 +1,93 @@
+
+    if (processedAttributes) {
+      const newImg = h('img', processedAttributes);
+      const parent = parents[parents.length - 1];
```
Since we are now using visit instead of visitParents so I removed this code and directly assigned new processed image to existing image using Object.assign()

tripodsan commented on src/steps/process-external-images.js:
```diff
@@ -0,0 +1,93 @@
+    return false;
+  }
+
+  // Check if URL matches any external image prefix
+  const { externalImageUrlPrefixes = [] } = state.config || {};
+  return externalImageUrlPrefixes.some((prefix) => src.startsWith(prefix));
```
what I mean is: since it's only possible to have external images in the markdown, if this was was enabled when ingesting via html2md, there is no need to check it here in the pipeline again.

gazal2708 replied:
> what I mean is: since it's only possible to have external images in the markdown, if this was enabled with the `externalImageUrlPrefixes` when ingesting via html2md, there is no need to check it here in the pipeline again.

but if any external/third party integration later happens in future (as mentioned by Raphael as well [Link to comment](https://github.com/adobe/helix-html-pipeline/pull/935#discussion_r2346988505) that doesnt use feature flag nor does it pushes images to media bus in html2md repo, then in that case, if we dont have the condition to handle only externalImageUrlPrefixes image url domains, we might end up processing those urls as well. What are your thoughts?

gazal2708 replied:
Okay, if thats the case, then let me revert this change.

gazal2708 replied:
I have reverted the change to -  check only for non media urls

gazal2708 committed:
Revert check for externalImageUrlPrefix

Commit:
Merge remote-tracking branch 'upstream/main'

gazal2708 committed:
added change for spreading existing properties along with new ones

Commit:
adding UT's for the external-images file to increase code coverage

tripodsan commented:
thanks @gazal2708 for you contribution, but it is very unlikely that we want to proceed with adding this feature. closing this PR for now.
